opensearch-operator/pkg/reconcilers/upgrade.go

```diff
@@ -0,0 +1,363 @@
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	r.osClient = clusterClient
+
+	//Fetch the working nodepool
+	nodePool, currentStatus := r.findWorkingNodePool()
+
+	// Work on the current nodepool as appropriate
+	switch currentStatus.Status {
+	case "Pending":
+		// Set it to upgrading and requeue
+		err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			if err := r.Get(r.ctx, client.ObjectKeyFromObject(r.instance), r.instance); err != nil {
+				return err
+			}
+			currentStatus.Status = "Upgrading"
+			r.instance.Status.ComponentsStatus = append(r.instance.Status.ComponentsStatus, currentStatus)
+			return r.Status().Update(r.ctx, r.instance)
+		})
+		r.recorder.Eventf(r.instance, "Normal", "upgrading", "beginning upgrade of data node pool %s", currentStatus.Component)
+		return ctrl.Result{
+			Requeue:      true,
+			RequeueAfter: 15 * time.Second,
+		}, err
+	case "NonDataPending":
+		// Set it to upgrading and requeue
+		err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			if err := r.Get(r.ctx, client.ObjectKeyFromObject(r.instance), r.instance); err != nil {
+				return err
+			}
+			currentStatus.Status = "UntrackedUpgrade"
+			r.instance.Status.ComponentsStatus = append(r.instance.Status.ComponentsStatus, currentStatus)
+			return r.Status().Update(r.ctx, r.instance)
+		})
+		r.recorder.Eventf(r.instance, "Normal", "upgrading", "beginning rollout of non data node pool %s", currentStatus.Component)
+		return ctrl.Result{
+			Requeue:      true,
+			RequeueAfter: 15 * time.Second,
+		}, err
+	case "Upgrading":
+		err := r.doDataNodeUpgrade(nodePool)
+		return ctrl.Result{
+			Requeue:      true,
+			RequeueAfter: 30 * time.Second,
+		}, err
+	case "Finished":
+		// Cleanup status after successful upgrade
+		err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			if err := r.Get(r.ctx, client.ObjectKeyFromObject(r.instance), r.instance); err != nil {
+				return err
+			}
+			r.instance.Status.Version = r.instance.Spec.General.Version
+			for _, pool := range r.instance.Spec.NodePools {
+				componentStatus := opsterv1.ComponentStatus{
+					Component:   "Upgrader",
+					Description: pool.Component,
+				}
+				currentStatus, found := helpers.FindFirstPartial(r.instance.Status.ComponentsStatus, componentStatus, helpers.GetByDescriptionAndGroup)
+				if found {
+					r.instance.Status.ComponentsStatus = helpers.RemoveIt(currentStatus, r.instance.Status.ComponentsStatus)
+				}
+			}
+			return r.Status().Update(r.ctx, r.instance)
+		})
+		r.recorder.Event(r.instance, "Normal", "upgrading", "cluster upgrade completed")
+		return ctrl.Result{}, err
+	default:
+		// We should never get here so return an error
+		return ctrl.Result{}, ErrUnexpectedStatus
+	}
+}
+
+// Currently provides basic validation on versions.
+// TODO Improve the validation (maybe allow patch version downgrades)
+func (r *UpgradeReconciler) validateUpgrade() error {
```

Comment:
that's run before the upgrade is starting ?

Reply:
Correcft